### ISSUE TITLE
AuthUserFile permissions check fails during SIGHUP, causing ProFTPD to stop

### ISSUE THREAD

**Issue by 2002demon** [proftpd/proftpd#1307]
### What I Did

After upgrade from debian 10 to 11 (1.3.6 to 1.3.7) reload does not work.
Using strace i see permission error to AuthUserFile passwd file => process died with TERM.
* `/panel` - `0710` `panel:panel`
* `/panel/proftpd `- `0755` `panel:panel`
* `/panel/proftpd/passwd` - `0600` `ftp:ftp`

Attention! 
User ftp is added to panel group, so normaly ftp user has access to `/panel/proftpd/passwd`
```
id ftp
uid=112(ftp) gid=65534(nogroup) groups=65534(nogroup),1000(panel)
```
I found fix for me and i use currently:
```
AuthFileOptions InsecurePerms
```

But I think there is a bug, because all permissions are ok.
Change /panel to 0755 helps too, but its not secure for me.

**Comment by Castaglia** [comment]
On restart, with permission checks of your `AuthUserFile` failing like this, ProFTPD will log a message saying _which_ permission is problematic.  Can you provide that log message, please?

**Comment by 2002demon** [comment]
**Restart is ok, but reload does not work:
/etc/init.d/proftpd reload**

2021-08-16 17:58:02,386 ns2.webserv24.com proftpd[1765815] 5.39.68.4: ProFTPD 1.3.7a (maint) (built Fri Jan 15 2021 14:09:32 UTC) standalone mode STARTUP
2021-08-16 17:58:21,899 ns2.webserv24.com proftpd[1765815] 5.39.68.4: received SIGHUP -- master server reparsing configuration file

**Nothing more in proftpd.log
But maybe strace help you:**

stat("/usr/lib/proftpd/mod_ifsession.so", {st_mode=S_IFREG|0644, st_size=27120, ...}) = 0
read(2, "", 4096)                       = 0
close(2)                                = 0
rt_sigprocmask(SIG_BLOCK, [HUP INT QUIT BUS USR1 ALRM TERM CHLD IO], NULL, 8) = 0
geteuid()                               = 0
setresgid(-1, 1005, -1)                 = 0
setresuid(-1, 113, -1)                  = 0
rt_sigprocmask(SIG_UNBLOCK, [HUP INT QUIT BUS USR1 ALRM TERM CHLD IO], NULL, 8) = 0
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 2
connect(2, {sa_family=AF_UNIX, sun_path="/var/run/nscd/socket"}, 110) = 0
sendto(2, "\2\0\0\0\0\0\0\0\4\0\0\0ftp\0", 16, MSG_NOSIGNAL, NULL, 0) = 16
poll([{fd=2, events=POLLIN|POLLERR|POLLHUP}], 1, 5000) = 1 ([{fd=2, revents=POLLIN|POLLHUP}])
read(2, "\2\0\0\0\1\0\0\0\4\0\0\0\2\0\0\0q\0\0\0\376\377\0\0\1\0\0\0\t\0\0\0"..., 36) = 36
read(2, "ftp\0x\0\0/srv/ftp\0/usr/sbin/nologi"..., 34) = 34
close(2)                                = 0
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 2
connect(2, {sa_family=AF_UNIX, sun_path="/var/run/nscd/socket"}, 110) = 0
sendto(2, "\2\0\0\0\2\0\0\0\4\0\0\0ftp\0", 16, MSG_NOSIGNAL, NULL, 0) = 16
poll([{fd=2, events=POLLIN|POLLERR|POLLHUP}], 1, 5000) = 1 ([{fd=2, revents=POLLIN|POLLHUP}])
read(2, "\2\0\0\0\1\0\0\0\4\0\0\0\2\0\0\0\355\3\0\0\0\0\0\0", 24) = 24
read(2, "ftp\0x\0", 6)                  = 6
close(2)                                = 0
lstat("/panel/proftpd/passwd", 0x7ffc4696aaf0) = -1 EACCES (Permission denied)
socket(AF_UNIX, SOCK_DGRAM, 0)          = 2
fcntl(2, F_SETFD, FD_CLOEXEC)           = 0
connect(2, {sa_family=AF_UNIX, sun_path="/dev/log"}, 110) = 0
getpid()                                = 1766094
sendto(2, "<28>Aug 16 18:01:22 proftpd[1766"..., 156, 0, NULL, 0) = 156
close(-1)                               = -1 EBADF (Bad file descriptor)
close(2)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++

**Comment by Castaglia** [comment]
This is the relevant line from that `strace` output:
```
lstat("/panel/proftpd/passwd", 0x7ffc4696aaf0) = -1 EACCES (Permission denied)
```
So the question then becomes: why can't ProFTPD read that path via `lstat(2)`?  To find out, we may need to see the output from these commands, please:
```
# ls -ald /panel /panel/proftpd
# ls -al /panel/proftpd/passwd
```

**Comment by 2002demon** [comment]
I have provided this info on the topic above ;)
Is the same: 
drwx--x--- 16 panel panel 4096 Aug 16 14:44 /panel
drwxr-xr-x  2 root  root  4096 Aug 11 09:19 /panel/proftpd
-rw------- 1 ftp ftp 152 Dec  2  2020 /panel/proftpd/passwd

If i change to ftp user, i have access to /panel/proftpd/passwd

**Comment by Castaglia** [comment]
You did indeed already provide that info; my mistake.

Would it also be possible get your full ProFTPD configuration (minus any passwords), to make sure that the configuration I'm using, to try to reproduce this behavior locally, matches what you're using?

**Comment by 2002demon** [comment]
Yes.
[proftpd.zip](https://github.com/proftpd/proftpd/files/7004834/proftpd.zip)

**Comment by 2002demon** [comment]
You need remove "AuthFileOptions InsecurePerms" to reproduce bug.

**Comment by Castaglia** [comment]
Thanks!  I'm now able to reproduce the issue locally.  When debug logging is enabled, and ProFTPD is restarted, these are the relevant log lines emitted:
```
2021-08-18 13:02:59,081 561e5e969772 proftpd[870]: mod_auth_file/1.0: unable to lstat AuthUserFile '/panel/proftpd/ftpd.passwd': Permission denied
2021-08-18 13:02:59,081 561e5e969772 proftpd[870]: fatal: AuthUserFile: unable to use /panel/proftpd/ftpd.passwd: Permission denied on line 25 of '/usr/local/etc/proftpd.conf'
```

**Comment by Castaglia** [comment]
For my reference, this bug occurs because root permissions are present when ProFTPD first starts up and parses its configuration, but are *not* present during a restart (when the possibly-changed configuration is parsed again).  And it appeared with the fix for [Bug #3892](http://bugs.proftpd.org/show_bug.cgi?id=3892), which added these permission checks.

**Comment by Castaglia** [comment]
This should now be fixed in master; I've backported the fix to the 1.3.7 branch as well.  Thanks!
